Automattic/woocommerce-payments
# Ensure subscription billing interval updates on first update

Fixes #3163

### Changes proposed in this Pull Request

This PR fixes a bug whereby updating a subscription product's billing interval requires two updates instead of one.

WCPay Subscriptions determines whether a product needs to be updated on Stripe by calling `maybe_schedule_product_create_or_update`, which checks the product's price data against the price data that's already in Stripe. This is currently hooked onto `woocommerce_update_product`. But when that hook is fired, the billing interval metadata hasn't updated yet, so we currently think there's nothing to update. Therefore, on the second update, we do notice that the billing data needs to be updated (since it was changed by the first update), and we schedule a product update.

As best I can tell, the reason why the price data isn't updated yet when we get the `woocommerce_update_product` hook is a result of [when subscription meta is updated by Subscriptions Core](https://github.com/Automattic/woocommerce-subscriptions-core/blob/21d95a568dc3e33b4e364d4a26abb82da41f1857/includes/admin/class-wc-subscriptions-admin.php#L85-L86):

```php
// Save subscription meta only when a subscription product is saved, can't run on the "'woocommerce_process_product_meta_' . $product_type" action because we need to override some WC defaults
add_action( 'save_post', __CLASS__ . '::save_subscription_meta', 11 );
```

Since subscription meta isn't added until a priority 11 hook on `save_post`, this PR addresses the bug by hooking onto `save_post` after the metadata is updated, so that when we call `maybe_schedule_product_create_or_update`, we have the updated price data and can catch the fact that an update is necessary.

#### Testing instructions

1. Go to **Products > Add Product** and add a new subscription product with subscription price $5 every month. Publish.
2. On Stripe, verify that you see the price of the subscription product as $5 per month.
3. Edit the product in WooCommerce, set the subscription price to $5 per week. Update.
4. On Stripe, the price should now be **correct** - it will now say $5 per week (before this PR, it would still say $5 per month).

-------------------


**Post merge**
